File: retizo_ai/android/app/src/main/kotlin/com/example/culai/EpsonDiscoveryHandler.kt
package com.example.culai

import android.content.Context
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.MethodChannel
import com.epson.epos2.discovery.Discovery
import com.epson.epos2.discovery.DiscoveryListener
import com.epson.epos2.discovery.FilterOption
import com.epson.epos2.discovery.DeviceInfo
import com.epson.epos2.Epos2Exception

/**
 * EpsonDiscoveryHandler - Manages printer discovery operations
 * 
 * Handles network/Bluetooth/USB printer discovery using Epson Discovery API
 * Provides callbacks to Flutter layer with discovered printer information
 * Supports filtering by port type and automatic timeout
 * 
 * Port Types:
 * - all: Discover all available printers
 * - tcp: Network/WiFi printers only
 * - bluetooth: Bluetooth printers only
 * - usb: USB-connected printers only
 * 
 * @param context Android context
 * @param channel Method channel for Flutter callbacks
 */
class EpsonDiscoveryHandler(
    private val context: Context,
    private val channel: MethodChannel
) {
    private var isDiscovering = false
    private val mainHandler = Handler(Looper.getMainLooper())
    private var timeoutRunnable: Runnable? = null
    
    companion object {
        private const val TAG = "EpsonDiscoveryHandler"
    }

    /**
     * Start printer discovery
     * 
     * @param portType Port filter: "all", "tcp", "bluetooth", "usb"
     * @param timeout Discovery timeout in milliseconds
     * @param result Method channel result callback
     */
    fun discoverPrinters(
        portType: String,
        timeout: Int,
        result: MethodChannel.Result
    ) {
        if (isDiscovering) {
            result.error("ALREADY_DISCOVERING", "Discovery already in progress", null)
            return
        }

        try {
            val filterOption = FilterOption()
            
            // Set port type filter
            when (portType.lowercase()) {
                "tcp" -> filterOption.portType = Discovery.PORTTYPE_TCP
                "bluetooth" -> filterOption.portType = Discovery.PORTTYPE_BLUETOOTH
                "usb" -> filterOption.portType = Discovery.PORTTYPE_USB
                else -> filterOption.portType = Discovery.PORTTYPE_ALL
            }
            
            // Set device type to printers only
            filterOption.deviceType = Discovery.TYPE_PRINTER
            
            // Use Epson name filter for better results
            filterOption.epsonFilter = Discovery.FILTER_NAME
            
            // Enable USB device name for better USB detection
            filterOption.usbDeviceName = Discovery.TRUE

            // Start discovery with listener
            Discovery.start(context, filterOption, discoveryListener)
            isDiscovering = true
            
            // Set timeout to auto-stop discovery
            timeoutRunnable = Runnable {
                if (isDiscovering) {
                    stopDiscoveryInternal()
                    notifyDiscoveryComplete()
                }
            }
            mainHandler.postDelayed(timeoutRunnable!!, timeout.toLong())
            
            result.success(mapOf(
                "status" to "started",
                "portType" to portType,
                "timeout" to timeout
            ))
            
        } catch (e: Exception) {
            android.util.Log.e(TAG, "Failed to start discovery", e)
            isDiscovering = false
            result.error("DISCOVERY_ERROR", e.message, getErrorDetails(e))
        }
    }

    /**
     * Stop ongoing printer discovery
     */
    fun stopDiscovery(result: MethodChannel.Result) {
        try {
            if (!isDiscovering) {
                // Allow callers to stop safely even if a timeout already fired
                result.success(mapOf("status" to "already_stopped"))
            } else {
                stopDiscoveryInternal()
                result.success(mapOf("status" to "stopped"))
            }
            
        } catch (e: Exception) {
            android.util.Log.e(TAG, "Failed to stop discovery", e)
            result.error("STOP_ERROR", e.message, getErrorDetails(e))
        }
    }

    /**
     * Internal method to stop discovery
     */
    private fun stopDiscoveryInternal() {
        try {
            Discovery.stop()
        } catch (e: Epos2Exception) {
            // Ignore ERR_PROCESSING - means discovery is already stopping
            if (e.errorStatus != Epos2Exception.ERR_PROCESSING) {
                android.util.Log.e(TAG, "Error stopping discovery", e)
            }
        } catch (e: Exception) {
            android.util.Log.e(TAG, "Unexpected error stopping discovery", e)
        }
        
        isDiscovering = false
        timeoutRunnable?.let { mainHandler.removeCallbacks(it) }
        timeoutRunnable = null
    }

    /**
     * Discovery listener - receives discovered printer information
     */
    private val discoveryListener = DiscoveryListener { deviceInfo ->
        try {
            // Extract printer information
            val printerData = mapOf(
                "deviceName" to deviceInfo.deviceName,
                "ipAddress" to (deviceInfo.ipAddress ?: ""),
                "macAddress" to (deviceInfo.macAddress ?: ""),
                "bdAddress" to (deviceInfo.bdAddress ?: ""),
                "target" to deviceInfo.target,
                "deviceType" to getDeviceTypeName(deviceInfo.deviceType),
                "printerSeries" to (deviceInfo.deviceName ?: "Unknown")
            )
            
            // Notify Flutter about discovered printer
            mainHandler.post {
                channel.invokeMethod("onPrinterDiscovered", printerData)
            }
            
            android.util.Log.d(TAG, "Discovered: ${deviceInfo.deviceName} at ${deviceInfo.target}")
            
        } catch (e: Exception) {
            android.util.Log.e(TAG, "Error processing discovered device", e)
        }
    }

    /**
     * Notify Flutter that discovery has completed
     */
    private fun notifyDiscoveryComplete() {
        mainHandler.post {
            channel.invokeMethod("onDiscoveryComplete", mapOf("status" to "completed"))
        }
    }

    /**
     * Get human-readable device type name
     */
    private fun getDeviceTypeName(deviceType: Int): String {
        return when (deviceType) {
            Discovery.TYPE_PRINTER -> "Printer"
            Discovery.TYPE_HYBRID_PRINTER -> "Hybrid Printer"
            Discovery.TYPE_DISPLAY -> "Display"
            Discovery.TYPE_KEYBOARD -> "Keyboard"
            Discovery.TYPE_SCANNER -> "Scanner"
            Discovery.TYPE_SERIAL -> "Serial Device"
            else -> "Unknown"
        }
    }

    /**
     * Get detailed error information for debugging
     */
    private fun getErrorDetails(e: Exception): Map<String, Any> {
        val details = mutableMapOf<String, Any>(
            "exception" to e.javaClass.simpleName,
            "message" to (e.message ?: "Unknown error")
        )
        
        if (e is Epos2Exception) {
            details["errorStatus"] = e.errorStatus
            details["errorCode"] = getEpos2ErrorName(e.errorStatus)
        }
        
        return details
    }

    /**
     * Get Epson error code name
     */
    private fun getEpos2ErrorName(errorStatus: Int): String {
        return when (errorStatus) {
            Epos2Exception.ERR_PARAM -> "ERR_PARAM"
            Epos2Exception.ERR_CONNECT -> "ERR_CONNECT"
            Epos2Exception.ERR_TIMEOUT -> "ERR_TIMEOUT"
            Epos2Exception.ERR_MEMORY -> "ERR_MEMORY"
            Epos2Exception.ERR_ILLEGAL -> "ERR_ILLEGAL"
            Epos2Exception.ERR_PROCESSING -> "ERR_PROCESSING"
            Epos2Exception.ERR_NOT_FOUND -> "ERR_NOT_FOUND"
            Epos2Exception.ERR_IN_USE -> "ERR_IN_USE"
            Epos2Exception.ERR_TYPE_INVALID -> "ERR_TYPE_INVALID"
            else -> "UNKNOWN_ERROR"
        }
    }

    /**
     * Cleanup resources
     */
    fun cleanup() {
        if (isDiscovering) {
            stopDiscoveryInternal()
        }
        timeoutRunnable?.let { mainHandler.removeCallbacks(it) }
    }
}

File: retizo_ai/ios/Runner/EpsonDiscoveryHandler.swift
import Foundation
import Flutter
import libepos2

// MARK: - EpsonDiscoveryHandler
//
// iOS mirror of EpsonDiscoveryHandler.kt.
// Uses the Epos2Discovery API — same semantic interface as the Android Discovery class.
// Discovered printers are forwarded to Flutter via the channel's invokeMethod,
// exactly matching the Android implementation's structure.

class EpsonDiscoveryHandler: NSObject {

    // MARK: - Private state

    private weak var channel: FlutterMethodChannel?
    private var isDiscovering = false
    private var timeoutWorkItem: DispatchWorkItem?
    private let mainQueue = DispatchQueue.main
    private let TAG = "EpsonDiscoveryHandler"

    // MARK: - Init

    init(channel: FlutterMethodChannel) {
        self.channel = channel
        super.init()
    }

    // MARK: - Start Discovery

    func discoverPrinters(portType: String,
                          timeout: Int,
                          result: @escaping FlutterResult) {
        if isDiscovering {
            result(FlutterError(code: "ALREADY_DISCOVERING",
                               message: "Discovery already in progress",
                               details: nil))
            return
        }

        let filterOption = Epos2FilterOption()

        // Port type filter — matches Android portType string mapping
        switch portType.lowercased() {
        case "tcp":
            filterOption.portType = EPOS2_PORTTYPE_TCP.rawValue
        case "bluetooth":
            filterOption.portType = EPOS2_PORTTYPE_BLUETOOTH.rawValue
        case "usb":
            filterOption.portType = EPOS2_PORTTYPE_USB.rawValue
        default:
            filterOption.portType = EPOS2_PORTTYPE_ALL.rawValue
        }

        filterOption.deviceType    = EPOS2_TYPE_PRINTER.rawValue

        let startResult = Epos2Discovery.start(filterOption, delegate: self)
        if startResult != EPOS2_SUCCESS.rawValue {
            result(FlutterError(code: "DISCOVERY_ERROR",
                               message: "Failed to start discovery: \(errorName(startResult))",
                               details: nil))
            return
        }

        isDiscovering = true

        // Auto-stop after timeout (mirrors Android mainHandler.postDelayed)
        let workItem = DispatchWorkItem { [weak self] in
            guard let self = self else { return }
            if self.isDiscovering {
                self.stopDiscoveryInternal()
                self.channel?.invokeMethod("onDiscoveryComplete",
                                           arguments: ["status": "completed"])
            }
        }
        timeoutWorkItem = workItem
        mainQueue.asyncAfter(deadline: .now() + .milliseconds(timeout), execute: workItem)

        NSLog("%@: discovery started — portType=%@, timeout=%dms", TAG, portType, timeout)
        result(["status": "started", "portType": portType, "timeout": timeout])
    }

    // MARK: - Stop Discovery

    func stopDiscovery(result: @escaping FlutterResult) {
        if !isDiscovering {
            result(["status": "already_stopped"])
            return
        }
        stopDiscoveryInternal()
        result(["status": "stopped"])
    }

    // MARK: - Internal Stop

    private func stopDiscoveryInternal() {
        let stopResult = Epos2Discovery.stop()
        // ERR_PROCESSING means already stopping — safe to ignore, matches Android behaviour
        if stopResult != EPOS2_SUCCESS.rawValue && stopResult != EPOS2_ERR_PROCESSING.rawValue {
            NSLog("%@: unexpected stop error: %@", TAG, errorName(stopResult))
        }
        isDiscovering = false
        timeoutWorkItem?.cancel()
        timeoutWorkItem = nil
    }

    // MARK: - Cleanup

    func cleanup() {
        if isDiscovering {
            stopDiscoveryInternal()
        }
    }

    // MARK: - Helpers

    private func errorName(_ code: Int32) -> String {
        switch code {
        case EPOS2_SUCCESS.rawValue:        return "SUCCESS"
        case EPOS2_ERR_PARAM.rawValue:      return "ERR_PARAM"
        case EPOS2_ERR_CONNECT.rawValue:    return "ERR_CONNECT"
        case EPOS2_ERR_TIMEOUT.rawValue:    return "ERR_TIMEOUT"
        case EPOS2_ERR_MEMORY.rawValue:     return "ERR_MEMORY"
        case EPOS2_ERR_ILLEGAL.rawValue:    return "ERR_ILLEGAL"
        case EPOS2_ERR_PROCESSING.rawValue: return "ERR_PROCESSING"
        case EPOS2_ERR_NOT_FOUND.rawValue:  return "ERR_NOT_FOUND"
        case EPOS2_ERR_IN_USE.rawValue:     return "ERR_IN_USE"
        case EPOS2_ERR_FAILURE.rawValue:    return "ERR_FAILURE"
        default:                            return "UNKNOWN_\(code)"
        }
    }

    private func deviceTypeName(_ type: Int32) -> String {
        switch type {
        case EPOS2_TYPE_PRINTER.rawValue:        return "Printer"
        case EPOS2_TYPE_HYBRID_PRINTER.rawValue: return "Hybrid Printer"
        case EPOS2_TYPE_DISPLAY.rawValue:        return "Display"
        case EPOS2_TYPE_KEYBOARD.rawValue:       return "Keyboard"
        case EPOS2_TYPE_SCANNER.rawValue:        return "Scanner"
        case EPOS2_TYPE_SERIAL.rawValue:         return "Serial Device"
        default:                                 return "Unknown"
        }
    }
}

// MARK: - Epos2DiscoveryDelegate

extension EpsonDiscoveryHandler: Epos2DiscoveryDelegate {
    func onDiscovery(_ deviceInfo: Epos2DeviceInfo!) {
        guard let info = deviceInfo else { return }

        let printerData: [String: Any] = [
            "deviceName":    info.deviceName ?? "",
            "ipAddress":     info.ipAddress ?? "",
            "macAddress":    info.macAddress ?? "",
            "bdAddress":     info.bdAddress ?? "",
            "target":        info.target ?? "",
            "deviceType":    deviceTypeName(info.deviceType),
            "printerSeries": info.deviceName ?? "Unknown"
        ]

        NSLog("%@: discovered: %@ at %@", TAG, info.deviceName ?? "?", info.target ?? "?")

        mainQueue.async { [weak self] in
            self?.channel?.invokeMethod("onPrinterDiscovered", arguments: printerData)
        }
    }
}
